### PR TITLE
Patterns types excerpts

### PR DESCRIPTION
**Pull request by MaryaBelanger**

### CODE REVIEW

**Comment by parlough** [comment]
Can you update this after your changes from https://github.com/dart-lang/site-www/commit/8d428ee46cd667080a6c56476633f5a6b658e884?

**Comment by MaryaBelanger** [comment]
@parlough it's building now, thanks for the heads up!

File: src/language/pattern-types.md [hunk 1]
```diff
@@ -142,6 +152,10 @@ A logical-or pattern separates subpatterns by `||` and matches if any of the
 branches match. Branches are evaluated left-to-right. Once a branch matches, the
 rest are not evaluated.
 
+Logical-or patterns are useful for having multiple cases share a body in [switch
+expressions or statements][].
```
Comment (parlough) [inline src/language/pattern-types.md:156]: It looks you introduced this during your merge, but as far as I can tell, this link isn't defined.

Broken link definitions will be caught after https://github.com/dart-lang/site-www/pull/4900 lands :)

File: examples/language/lib/patterns/pattern_types.dart [hunk 2]
```diff
@@ -0,0 +1,262 @@
...
+      return switch (char) {
+        < space => 'control',
+        == space => 'space',
+        > space && < zero => 'punctuation',
+        >= zero && <= nine => 'digit',
+        _ => ''
+      };
+    }
+    // #enddocregion relational
+  }
```
Comment (parlough) [inline examples/language/lib/patterns/pattern_types.dart:225]: We should eventually add tests for cases like these, but for now, let's add an assert to make the function used.

```suggestion
    assert(asciiCharType(32) == 'space');
  }
```

File: examples/language/lib/patterns/pattern_types.dart [hunk 3]
```diff
@@ -0,0 +1,262 @@
+// ignore_for_file: duplicate_variable_pattern, duplicate_definition, unused_local_variable, dead_code, constant_pattern_never_matches_value_type, unnecessary_null_check_pattern, pattern_never_matches_value_type, unreachable_switch_case, unused_element
```
Comment (parlough) [inline examples/language/lib/patterns/pattern_types.dart:1]: Let's split these up to make them easier to read and edit. Preferably we can remove more of them over time. This already removes `unused_element`.

```suggestion
// ignore_for_file: duplicate_variable_pattern, duplicate_definition, unused_local_variable
// ignore_for_file: dead_code, constant_pattern_never_matches_value_type, unnecessary_null_check_pattern
// ignore_for_file: pattern_never_matches_value_type, unreachable_switch_case
```

File: examples/language/lib/patterns/pattern_types.dart [hunk 4]
```diff
@@ -0,0 +1,262 @@
...
+
+class Color {
+  static const red = true;
+  static const yellow = false;
+  static const blue = true;
+}
+
+class Rect {
+  int width;
+  int height;
```
Comment (parlough) [inline examples/language/lib/patterns/pattern_types.dart:23]: Prefer defaulting to `final` for fields to catch mistakes later on.

```suggestion
  final int width;
  final int height;
```

File: examples/language/lib/patterns/pattern_types.dart [hunk 5]
```diff
@@ -0,0 +1,262 @@
...
+  Rect({required this.width, required this.height});
+}
+
+class Point {
+  int? x;
+  int? y;
```
Comment (parlough) [inline examples/language/lib/patterns/pattern_types.dart:29]: Prefer defaulting to `final` for fields to catch mistakes later on.

```suggestion
  final int? x;
  final int? y;
```

**Review by parlough — APPROVED** [review]
Thanks for this, just a few minor adjustments. It's generally better to avoid as many `ignore_for_file`s as possible, since they can potentially hide issues later on. That's fine for now, as we can go back and make improvements later to avoid the lints, or make the ignores more local.

File: src/language/pattern-types.md (hunk 1 above)
Comment (MaryaBelanger) [reply src/language/pattern-types.md:156]: Ah that's left over somehow from when I moved that comment to the patterns page. Thanks for catching!

**Comment by parlough** [comment]
Yeah not sure why it doesn't say formatting? I'll see if we can make it more clear :P

**Pull request merged**